File: src/main/java/lesson10/Tunnel.java
package lesson10;

import java.util.concurrent.Semaphore;

public class Tunnel extends Stage {
    public Tunnel() {
        this.length = 80;
        this.description = "Тоннель " + length + " метров";
    }

    public static final int COMPETITORS_COUNT_IN_TUNNEL = 2;

    private Semaphore semaphore = new Semaphore(COMPETITORS_COUNT_IN_TUNNEL);

    @Override
    public void overcome(Car c) {
        try {
            overcoveInternal(c);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void overcoveInternal(Car c) throws InterruptedException {
        System.out.println(c.getName() + " готовится к этапу(ждет): " + description + "     " + System.currentTimeMillis());
        semaphore.acquire();
        System.out.println(c.getName() + " начал этап: " + description + "     " + System.currentTimeMillis());
        Thread.sleep(length / c.getSpeed() * 1000);
        System.out.println(c.getName() + " закончил этап: " + description + "     " + System.currentTimeMillis());
        semaphore.release();
    }
}

File: src/main/kotlin/lesson10_kt/Tunnel.kt
package lesson10_kt

import java.util.concurrent.Semaphore

class Tunnel : Stage() {
    init {
        this.length = 80
        this.description = "Тоннель $length метров"
    }

    val semaphore =  Semaphore(2)

    override fun overcome(c: Car) {
        try {
            try {
                println(c.name + " готовится к этапу(ждет): " + description)
                semaphore.acquire()
                println(c.name + " начал этап: " + description)
                Thread.sleep((length / c.speed * 1000).toLong())
            } catch (e: InterruptedException) {
                e.printStackTrace()
            } finally {
                println(c.name + " закончил этап: " + description)
                semaphore.release()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}